### PR TITLE
Store

### PR DESCRIPTION
**Pull request by xWafl**
You can purchase roles with codify coins.

Features:
* Purchase roles with one-time subscription or every x days
* Members with the "manage roles" permission can add items with `cc!addStoreItem`
* You can see the store for a server with `cc!store`
* Store items are by store, each server has a different store.
* You can unsubscribe from an item with `cc!unsubscribe`
* View your purchases with `cc!purchases`

### CODE REVIEW

File: src/modules/commands/unsubscribe.ts [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+import knex from "../../../db/knex";
+import Subscription from "../../common/types/Subscription";
+import Store from "../../common/types/Store";
+
+export default new Command()
+    .match(matchPrefixesStrict("unsubscribe"))
+    .use<ParseArgumentsState>(async context => {
+        const { message } = context;
+        const { args } = context.state;
+
```
Comment (Vyctor661) [inline src/modules/commands/unsubscribe.ts:14]: what happens if there are no arguments

File: src/modules/commands/buy.ts [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+
+    return parseInt(balance) >= amount;
+};
+
+export default new Command()
+    .match(matchPrefixesStrict("buy"))
+    .use<ParseArgumentsState>(async context => {
+        const { message } = context;
+        const { args } = context.state;
+
```
Comment (Vyctor661) [inline src/modules/commands/buy.ts:24]: what happens if there are no arguments

File: src/modules/commands/addStoreItem.ts [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+import { matchPrefixesStrict } from "../../common/matching/matchPrefixesStrict";
+import knex from "../../../db/knex";
+import Store from "../../common/types/Store";
+
+export default new Command()
+    .match(matchPrefixesStrict("addStoreItem"))
+    .use<ParseArgumentsState>(async context => {
+        const { message } = context;
+        const { args } = context.state;
+
```
Comment (Vyctor661) [inline src/modules/commands/addStoreItem.ts:13]: what happens if there are no arguments

**Review by Vyctor661 — CHANGES_REQUESTED** [review]

File: src/common/knexCommon.ts [hunk 4]
```diff
@@ -83,3 +85,66 @@ export async function autoXpClaim(userid: string, ctx: Message) {
...
+                    await knex("user")
+                        .update({
+                            balance: dbUser.balance - store.price
+                        })
+                        .where({ userid: l.userId });
+
+                    await knex<Subscription>("subscriptions")
+                        .update({
+                            expiration: (
+                                Math.floor(Date.now() / 1000) + 86400
```
[the remark this answers is not in the document]
Comment (xWafl) [reply src/common/knexCommon.ts:142]: I was going to type my reasoning but I think I will just add something that bills you multiple times in 1 message if you spend longer than the interval away from the server, which will fix the bug.

File: src/modules/commands/addStoreItem.ts [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+            );
+        }
+        const subscriptionInterval = args[2] && Number(args[2]);
+
+        await knex<Store>("store").insert({
+            serverId: message.guild.id,
+            roleId,
+            price,
+            subscription: !!subscriptionInterval,
+            subscriptionInterval: subscriptionInterval || 0
```
[the remark this answers is not in the document]
Comment (xWafl) [reply src/modules/commands/addStoreItem.ts:43]: Huh?

File: src/modules/commands/buy.ts [hunk 6]
```diff
@@ -0,0 +1,85 @@
+import { Command } from "@enitoni/gears-discordjs";
+
+import { ParseArgumentsState } from "../../common/parsing/middleware/parseArguments";
+import { matchPrefixesStrict } from "../../common/matching/matchPrefixesStrict";
+import knex from "../../../db/knex";
+import Store from "../../common/types/Store";
+import Subscription from "../../common/types/Subscription";
+
+const checkBalance = async (amount: number, id: string) => {
+    if (amount <= 0 || amount >= 1001) {
```
[the remark this answers is not in the document]
Comment (xWafl) [reply src/modules/commands/buy.ts:10]: Oh, oops

File: src/modules/commands/addStoreItem.ts [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+                ":x: **Oops,** you aren't allowed to do that. Make sure you have the `Manage roles` permission."
+            );
+        }
+        if (!message.guild.roles.get(args[0])) {
+            return message.channel.send(
+                ":x: **Oops,** looks like that role doesn't exist."
+            );
+        }
+        const roleId = args[0];
+        if (Number(args[1]).toString() !== args[1]) {
```
[the remark this answers is not in the document]
Comment (xWafl) [reply src/modules/commands/addStoreItem.ts:25]: I should add a check for that

File: src/index.ts [hunk 8]
```diff
@@ -26,6 +26,11 @@ import createWebhook from "./modules/commands/createWebhook";
 import addQuote from "./modules/commands/addquote";
 import disableLevelupMessages from "./modules/commands/disableLevelupMessages";
 import trivia from "./modules/commands/trivia";
+import store from "./modules/commands/store";
+import addStoreItem from "./modules/commands/addStoreItem";
```
[the remark this answers is not in the document]
Comment (xWafl) [reply src/index.ts:30]: True, didn't think of that.

File: src/modules/commands/unsubscribe.ts (hunk 1 above)
Comment (xWafl) [reply src/modules/commands/unsubscribe.ts:14]: Forgot about that

File: src/modules/commands/buy.ts (hunk 2 above)
Comment (xWafl) [reply src/modules/commands/buy.ts:24]: Forgot about that

File: src/modules/commands/addStoreItem.ts (hunk 3 above)
Comment (xWafl) [reply src/modules/commands/addStoreItem.ts:13]: Again, forgot about that

**Pull request merged**